File: ejercicios/fase-1/capstone-misma-app-dos-lenguajes/python/pantry.py
"""Lógica de dominio de la despensa (sin HTTP).

Completa los TODO. NO cambies las firmas de los métodos públicos: los tests de
`test_pantry.py` dependen de ellas.
"""

from __future__ import annotations

import json
from pathlib import Path

from pydantic import BaseModel, Field, ValidationError  # noqa: F401  (lo usarás)


# TODO (1.4): define el modelo de un ítem NUEVO (sin id) con validación declarativa:
#   - name: texto NO vacío
#   - quantity: número > 0
#   - unit: texto NO vacío
# Pista: usa Field(min_length=1) para texto y Field(gt=0) para el número.
class NewItem(BaseModel):
    ...  # reemplaza esto por los campos validados


# TODO: un ítem ya guardado ES un NewItem + un id entero.
class Item(NewItem):
    ...  # reemplaza esto


class PantryStore:
    """Guarda ítems en un archivo JSON. Recibe el path por parámetro (seam de test)."""

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)
        # TODO (1.5): si el archivo no existe, créalo con una lista JSON vacía ("[]").
        raise NotImplementedError("implementa el constructor")

    def list_items(self) -> list[dict]:
        # TODO: lee el archivo JSON y devuelve la lista de ítems.
        raise NotImplementedError

    def add_item(self, data: dict) -> dict:
        # TODO:
        #   1) valida `data` con NewItem (deja que ValidationError suba si es inválido),
        #   2) calcula el próximo id = max(ids existentes, default 0) + 1,
        #   3) construye el Item, persístelo y devuélvelo como dict.
        raise NotImplementedError

    def get_item(self, item_id: int) -> dict | None:
        # TODO: devuelve el ítem con ese id, o None si no existe.
        raise NotImplementedError

    def remove_item(self, item_id: int) -> bool:
        # TODO: borra el ítem con ese id. Devuelve True si borró, False si no existía.
        raise NotImplementedError

File: ejercicios/fase-1/capstone-misma-app-dos-lenguajes/typescript/pantry.ts
/**
 * Lógica de dominio de la despensa (sin HTTP).
 *
 * Completa los TODO. NO cambies las firmas públicas: los tests de `pantry.test.ts`
 * dependen de ellas.
 */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { z } from "zod";

// TODO (1.8): define el esquema zod de un ítem NUEVO (sin id):
//   - name: string NO vacío        -> z.string().min(1)
//   - quantity: número > 0          -> z.number().positive()
//   - unit: string NO vacío         -> z.string().min(1)
export const NewItemSchema = z.object({
  // ...completa los campos
});

// `z.infer` deriva el tipo TypeScript del esquema: no repites la forma dos veces.
export type NewItem = z.infer<typeof NewItemSchema>;
export type Item = NewItem & { id: number };

export class PantryStore {
  constructor(private readonly path: string) {
    // TODO (1.5): si el archivo no existe, créalo con "[]".
    throw new Error("TODO: implementa el constructor");
  }

  list(): Item[] {
    // TODO: lee el archivo JSON y devuelve la lista de ítems.
    throw new Error("TODO: list");
  }

  add(data: unknown): Item {
    // TODO:
    //   1) valida `data` con NewItemSchema.parse(data) (lanza ZodError si no valida),
    //   2) calcula el próximo id = max(ids, 0) + 1,
    //   3) construye el Item, persístelo y devuélvelo.
    throw new Error("TODO: add");
  }

  get(id: number): Item | undefined {
    // TODO: devuelve el ítem con ese id, o undefined si no existe.
    throw new Error("TODO: get");
  }

  remove(id: number): boolean {
    // TODO: borra el ítem con ese id. Devuelve true si borró, false si no existía.
    throw new Error("TODO: remove");
  }
}
